x-pack/plugins/apm/public/components/app/error_group_details/error_sampler/error_sample_contextual_insight.tsx

```diff
@@ -0,0 +1,125 @@
+import { i18n } from '@kbn/i18n';
+import {
+  useObservabilityAIAssistant,
+  ContextualInsight,
+  type Message,
+  MessageRole,
+} from '@kbn/observability-ai-assistant-plugin/public';
+import React, { useMemo, useState } from 'react';
+import { APMError } from '../../../../../typings/es_schemas/ui/apm_error';
+import { Transaction } from '../../../../../typings/es_schemas/ui/transaction';
+import { ErrorSampleDetailTabContent } from './error_sample_detail';
+import { exceptionStacktraceTab, logStacktraceTab } from './error_tabs';
+
+export function ErrorSampleContextualInsight({
+  error,
+  transaction,
+}: {
+  error: APMError;
+  transaction?: Transaction;
+}) {
+  const aiAssistant = useObservabilityAIAssistant();
+
+  const [logStacktrace, setLogStacktrace] = useState('');
+  const [exceptionStacktrace, setExceptionStacktrace] = useState('');
+
+  const messages = useMemo<Message[]>(() => {
+    const now = new Date().toISOString();
+
+    const serviceName = error.service.name;
+    const languageName = error.service.language?.name ?? '';
+    const runtimeName = error.service.runtime?.name ?? '';
+    const runtimeVersion = error.service.runtime?.version ?? '';
+    const transactionName = transaction?.transaction.name ?? '';
+
+    return [
+      {
+        '@timestamp': now,
+        message: {
+          role: MessageRole.System,
+          content: `You are apm-gpt, a helpful assistant for performance analysis, optimisation and
+          root cause analysis of software. Answer as concisely as possible.`,
+        },
+      },
+      {
+        '@timestamp': now,
+        message: {
+          role: MessageRole.User,
+          content: `I'm an SRE. I am looking at an exception and trying to understand what it means.
+
+          Your task is to describe what the error means and what it could be caused by.
+    
+          The error occurred on a service called ${serviceName}, which is a ${runtimeName} service written in ${languageName}. The
+          runtime version is ${runtimeVersion}.
+    
+          The request it occurred for is called ${transactionName}.
+    
+          ${
+            logStacktrace
+              ? `The log stacktrace:
+          ${logStacktrace}`
+              : ''
+          }
+    
+          ${
+            exceptionStacktrace
+              ? `The exception stacktrace:
+          ${exceptionStacktrace}`
+              : ''
+          }
+          `,
+        },
+      },
+    ];
+  }, [error, transaction, logStacktrace, exceptionStacktrace]);
+
+  return aiAssistant.isEnabled() && messages ? (
+    <>
+      <EuiFlexItem>
+        <ContextualInsight
+          messages={messages}
+          title={i18n.translate(
+            'xpack.apm.errorGroupContextualInsight.explainErrorTitle',
+            { defaultMessage: "What's this error?" }
+          )}
+        />
+      </EuiFlexItem>
+      <EuiSpacer size="s" />
+      <div
+        ref={(next) => {
```

Comment:
Really loved the idea of having hidden div with a ref feeding the rendered child to ContextualInsights

Reply:
I'm going to be honest, I don't 😂 I would prefer it if we parse the exception to something that we can both render to the DOM and to text, but it's hard to convert what we have today (which are lots of different components) to something that returns structural data.

Reply:
I too initially started thinking in the same direction as to why this round and round. But then after looking at the exception and logs rendering logic, i was like, naa given what we have, this is really good